consular/clients.py

```diff
@@ -0,0 +1,287 @@
+
+import treq
+
+
+class JsonClient(object):
+    debug = False
+    clock = reactor
+    timeout = 5
+    requester = lambda self, *a, **kw: treq.request(*a, **kw)
+
+    def __init__(self, endpoint):
+        """
+        Create a client with the specified default endpoint.
+        """
+        self.endpoint = endpoint
+        self.pool = client.HTTPConnectionPool(self.clock, persistent=False)
+
+    def _log_http_response(self, response, method, path, data):
+        log.msg('%s %s with %s returned: %s' % (
+            method, path, data, response.code))
+        return response
+
+    def _log_http_error(self, failure, url):
+        log.err(failure, 'Error performing request to %s' % (url,))
+        return failure
+
+    def request(self, method, path, endpoint=None, json_data=None, **kwargs):
+        """
+        Perform a request. A number of basic defaults are set on the request
+        that make using a JSON API easier. These defaults can be overridden by
+        setting the parameters in the keyword args.
+
+        :param: method:
+            The HTTP method to use (example is `GET`).
+        :param: path:
+            The URL path. This is appended to the endpoint and should start
+            with a '/' (example is `/v2/apps`).
+        :param: endpoint:
+            The URL endpoint to use. The default value is the endpoint this
+            client was created with (`self.endpoint`) (example is
+            `http://localhost:8080`)
+        :param: json_data:
+            A python data structure that will be converted to a JSON string
+            using `json.dumps` and used as the request body.
+        :param: kwargs:
+            Any other parameters that will be passed to `treq.request`, for
+            example headers or parameters.
+        """
+        url = ('%s%s' % (endpoint or self.endpoint, path)).encode('utf-8')
+
+        data = json.dumps(json_data) if json_data else None
+        requester_kwargs = {
+            'headers': {
+                'Content-Type': 'application/json',
+                'Accept': 'application/json',
+            },
+            'data': data,
+            'pool': self.pool,
+            'timeout': self.timeout
+        }
+        requester_kwargs.update(kwargs)
+
+        d = self.requester(method, url, **requester_kwargs)
+
+        if self.debug:
+            d.addCallback(self._log_http_response, method, url, data)
+
+        return d.addErrback(self._log_http_error, url)
+
+    def get_json(self, path, **kwargs):
+        """
+        Perform a GET request to the given path and return the JSON response.
+        """
+        d = self.request('GET', path, **kwargs)
+        return d.addCallback(self._response_json_if_ok)
+
+    def _response_json_if_ok(self, response):
+        """
+        Get the response JSON content if the respones code is OK (200), else
+        raise an `UnexpectedResponseError`.
+        """
+        if response.code == OK:
```

Comment:
It doesn't look like it http://twistedmatrix.com/trac/browser/tags/releases/twisted-15.4.0/twisted/web/_responses.py